File: conformance/tests/assert-false-guard/AssertFalseGuard.runar.rs
use runar::prelude::*;

/// AssertFalseGuard -- the `assert!(false)`-else guard, in the two positions
/// the multi-result branch node originally missed. `bump` is a single property
/// written under a guard whose else is the dead abort -- recognised as a
/// ONE-branch chain, so excluded from declaring its result, and not rewritten
/// either because the lift only rewrites chains of two or more. `dispatch` is
/// `Selector`'s exact chain one loop deeper, where the lift never walks.
#[runar::contract]
pub struct AssertFalseGuard {
    pub count: Bigint,
    pub a: Bigint,
    pub b: Bigint,
}

impl AssertFalseGuard {
    pub fn bump(&mut self, n: Bigint) {
        if n > 0 {
            self.count = self.count + n;
        } else {
            assert!(false);
        }
    }

    pub fn dispatch(&mut self, sel: Bigint, v: Bigint) {
        for i in 0..2 {
            if sel == 0 {
                self.a = v;
            } else {
                if sel == 1 {
                    self.b = v;
                } else {
                    assert!(false);
                }
            }
        }
    }
}

File: conformance/tests/assert-false-guard/AssertFalseGuard.runar.go
package x

import runar "github.com/icellan/runar/packages/runar-go"

// AssertFalseGuard -- the runar.Assert(false)-else guard, in the two positions
// the multi-result branch node originally missed. Bump is a single property
// written under a guard whose else is the dead abort -- recognised as a
// ONE-branch chain, so excluded from declaring its result, and not rewritten
// either because the lift only rewrites chains of two or more. Dispatch is
// Selector's exact chain one loop deeper, where the lift never walks.
type AssertFalseGuard struct {
	runar.StatefulSmartContract

	Count runar.Bigint
	A     runar.Bigint
	B     runar.Bigint
}

func (c *AssertFalseGuard) Bump(n runar.Bigint) {
	if n > 0 {
		c.Count = c.Count + n
	} else {
		runar.Assert(false)
	}
}

func (c *AssertFalseGuard) Dispatch(sel runar.Bigint, v runar.Bigint) {
	for i := runar.Bigint(0); i < 2; i++ {
		if sel == 0 {
			c.A = v
		} else if sel == 1 {
			c.B = v
		} else {
			runar.Assert(false)
		}
	}
}
